### PR TITLE
feat(api): upload docs to s3

### PR DESCRIPTION
**Pull request by Orta21**
Ref. metriport/metriport-internal#521

Ref: 521

### Dependencies

- Downstream: https://github.com/metriport/metriport-internal/pull/580

### Description

WIP upload docs Job to “Get All Documents”

### Release Plan

- Need to deploy the npm package 
- Also need to deploy the infra with new s3 bucket
- Run FHIR migration

- [ ] Merge once approved

### CODE REVIEW

File: api/app/src/external/commonwell/document/shared.ts [hunk 1]
```diff
@@ -1,39 +1,86 @@
...
+    masterIdentifier: {
+      system: doc.content?.masterIdentifier?.system,
+      value: doc.content?.masterIdentifier?.value,
+    },
+    identifier: doc.content?.identifier?.map(id => {
+      return {
+        system: id.system,
+        value: id.value,
+        // question
+        // use: id.use,
```
Comment (Orta21) [inline api/app/src/external/commonwell/document/shared.ts:43]: Medplum has an enum for use but our sdk schema does not. Should we update our sdk since medplum is using the FHIR specs anyways

File: api/app/src/external/commonwell/document/document-query.ts [hunk 2]
```diff
@@ -90,23 +101,101 @@ async function internalGetDocuments({
...
+
+  s3Refs.forEach(ref => {
+    if (ref.status === "rejected") {
+      capture.error(ref.reason, {
+        extra: {
+          context: `s3.documentUpload`,
+        },
+      });
+    }
+  });
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/document-query.ts:177]: Could this get crazy if 100s of documents fail? Should we bundle these up in a single capture?

@leite08 - any thoughts here would be good, I'm not familiar with how Sentry handles these situations

File: api/app/src/external/commonwell/document/document-query.ts [hunk 3]
```diff
@@ -1,40 +1,50 @@
...
+    }
+
+    const docsS3Refs = await dowloadAnduploadDocsToS3({
+      patient,
+      facilityId,
+      documents: cwDocuments,
     });
 
-    return await createOrUpdate(patient, documents);
+    return await sendToFHIR({ organization, patient, docs: docsS3Refs });
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/document-query.ts:47]: Would be better to only update the org/patient in the FHIR storage on org/patient create/update operations, ie on the org/patient endpoints only.

File: api/app/src/external/commonwell/document/shared.ts [hunk 4]
```diff
@@ -1,39 +1,86 @@
...
+              valueString: "S3",
+            },
+          ],
+        },
+      },
+    ],
+    extension: [
+      {
+        valueReference: {
+          reference: "Commonwell",
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/shared.ts:77]: nit: CommonWell -> we should probably define this string in a const somewhere, so we can use it throughout the application

File: api/app/src/external/commonwell/document/shared.ts [hunk 5]
```diff
@@ -1,39 +1,86 @@
...
+    content: [
+      {
+        attachment: {
+          title: doc.fileName,
+          contentType: "application/xml",
+          url: doc.content?.location,
+          extension: [
+            {
+              valueString: "S3",
+            },
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/shared.ts:69]: why is this necessary?

File: api/app/src/external/commonwell/document/shared.ts [hunk 6]
```diff
@@ -1,39 +1,86 @@
...
       },
-      raw: doc.raw,
-    };
+    ],
+    description: doc.content?.description,
+    content: [
+      {
+        attachment: {
+          title: doc.fileName,
+          contentType: "application/xml",
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/shared.ts:64]: is this always true? Not sure we can rely on this being xml 100% of the time

File: api/app/src/routes/medical/dtos/documentDTO.ts [hunk 7]
```diff
@@ -1,30 +1,58 @@
...
   fileName: string;
   location: string;
   description: string | undefined;
   status: string | undefined;
-  indexed: string | undefined; // ISO-8601
-  mimeType: string | undefined;
-  size: number | undefined; // bytes
+  indexed?: string | undefined; // ISO-8601
+  mimeType?: string | undefined;
+  size?: number | undefined; // bytes
```
Comment (Goncharo) [inline api/app/src/routes/medical/dtos/documentDTO.ts:14]: OOC - why are these optional now?

File: infra/lib/api-stack.ts [hunk 8]
```diff
@@ -282,6 +286,20 @@ export class APIStack extends Stack {
       uri: `http://${apiServerAddress}/{proxy}`,
     });
 
+    //-------------------------------------------
+    // S3 bucket for MAPI Access Requests
```
Comment (Goncharo) [inline infra/lib/api-stack.ts:290]: nit: comment out of date

File: api/app/src/routes/medical/document.ts [hunk 9]
```diff
@@ -65,30 +68,21 @@ router.post(
  *
  * Downloads the specified document for the specified patient.
```
Comment (Goncharo) [inline api/app/src/routes/medical/document.ts:69]: nit: comment is out of date

File: api/app/src/routes/medical/document.ts [hunk 10]
```diff
@@ -65,30 +68,21 @@ router.post(
...
- * @param req.query.facilityId The facility providing NPI for the document download.
- * @param req.query.location The document URL.
- * @param [req.query.mimeType] The mime type of the document.
- * @param [req.query.fileName] The file name of the document.
- * @return The document content.
+ * @param req.query.fileName The file name of the document in s3.
+ * @return presigned url
  */
 router.get(
   "/download",
```
Comment (Goncharo) [inline api/app/src/routes/medical/document.ts:75]: should this now be called `downloadUrl` or something?

File: api/app/src/routes/medical/document.ts [hunk 11]
```diff
@@ -65,30 +68,21 @@ router.post(
...
 
-    fileName && res.attachment(fileName);
-    mimeType && res.header("Content-Type", mimeType);
-
-    await downloadDocument({ patientId, cxId, facilityId, location, stream: res });
+    const url = await s3client.getSignedUrl("getObject", {
+      Bucket: Config.getMedicalDocumentsBucketName(),
+      Key: fileName,
+      Expires: 100,
+    });
```
Comment (Goncharo) [inline api/app/src/routes/medical/document.ts:83]: this seems like a security issue -> with this approach AFAIK there is nothing stopping a cx guessing file names, and downloading another cx's document

we will need to make this somehow tied to the individual cx only

File: packages/packages/api/src/medical/client/metriport.ts [hunk 12]
```diff
@@ -328,27 +328,20 @@ export class MetriportMedicalApi {
...
-   * @param facilityId The facility providing the NPI to support this operation.
-   * @param location The location of the document.
-   * @return The document's contents (bytes).
+   * @param fileName The file name of the document in s3.
+   * @return presigned url
    */
   //eslint-disable-next-line @typescript-eslint/no-explicit-any
-  async getDocument(patientId: string, facilityId: string, location: string): Promise<any> {
+  async getDocument(fileName: string): Promise<{ url: string }> {
     const resp = await this.api.get(`${DOCUMENT_URL}/download`, {
```
Comment (Goncharo) [inline packages/packages/api/src/medical/client/metriport.ts:339]: perhaps this can be `getDocumentUrl` & `/downloadUrl` now?

also, the docs will need updating when we agree on an approach

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: api/app/src/external/commonwell/document/document-query.ts (hunk 2 above)
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:177]: I can count them and then do a generic message so  its like 

extra: {
      context: `s3.documentUpload`,
      count: 12
}

File: api/app/src/external/commonwell/document/shared.ts (hunk 5 above)
Comment (Orta21) [reply api/app/src/external/commonwell/document/shared.ts:69]: I was thinking that we might have docs with different locations but thinking about it were only going to be storing the s3 docs we have saved in FHIR

File: packages/packages/api/src/medical/client/metriport.ts (hunk 12 above)
Comment (Orta21) [reply packages/packages/api/src/medical/client/metriport.ts:339]: This makes sense to me

File: api/app/src/routes/medical/dtos/documentDTO.ts (hunk 7 above)
Comment (Orta21) [reply api/app/src/routes/medical/dtos/documentDTO.ts:14]: Fixed

File: api/app/src/routes/medical/document.ts [hunk 13]
```diff
@@ -63,32 +67,25 @@ router.post(
...
- * @param req.query.location The document URL.
- * @param [req.query.mimeType] The mime type of the document.
- * @param [req.query.fileName] The file name of the document.
- * @return The document content.
+ * @param req.query.fileName The file name of the document in s3.
+ * @return presigned url
  */
 router.get(
-  "/download",
+  "/downloadUrl",
```
Comment (leite08) [inline api/app/src/routes/medical/document.ts:76]: Didn't we want to keep the URL abstracted away from the caller, by keeping this `/download` and returning a redirect/302 to the temp signed S3 URL?

File: api/app/src/routes/medical/document.ts [hunk 14]
```diff
@@ -63,32 +67,25 @@ router.post(
...
-    fileName && res.attachment(fileName);
-    mimeType && res.header("Content-Type", mimeType);
+    const hour = 3600;
+    const fileName = getFromQueryOrFail("fileName", req);
 
-    await downloadDocument({ patientId, cxId, facilityId, location, stream: res });
+    const url = await s3client.getSignedUrl("getObject", {
+      Bucket: Config.getMedicalDocumentsBucketName(),
+      Key: createS3FileName(cxId, fileName),
+      Expires: hour,
```
Comment (leite08) [inline api/app/src/routes/medical/document.ts:85]: I think this is too long. A pre-signed URL is basically open to anyone with access to it (I'd say this is one of the reasons CW does not return a URL but has the expense of transferring all files that go through the network using their servers).

That's another reason why we should not return the URL, but send a redirect to a much shorter lived pre-signed URL (maybe 10-20 seconds).

File: utils/src/fhir-migration.ts [hunk 15]
```diff
@@ -0,0 +1,200 @@
+import * as dotenv from "dotenv";
```
Comment (Orta21) [inline utils/src/fhir-migration.ts:1]: I figured this file we would only run once so i copied and pasted logic that we had in the api into here to save time

File: api/app/src/external/fhir/organization/upsert-organization.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+import { capture } from "../../../shared/notifications";
+
+export const upsertOrgToFHIRServer = async (organization: Organization) => {
+  try {
+    return await api.updateResource(organization);
+  } catch (err) {
+    capture.error(err, {
+      extra: { context: `fhir.add.organization` },
+    });
+  }
```
Comment (leite08) [inline api/app/src/external/fhir/organization/upsert-organization.ts:12]: I see that in this case this is an additional call to store the Org on the FHIR server - beyond doing that on Postgres, so I think its nice to have an error handling to indicate a potential error is specific with the FHIR server upsert, not our BE's. (this shouldn't be needed if we had proper source maps on Sentry, though)

But I think we should be rethrowing errors, no?

File: api/app/src/routes/medical/document.ts [hunk 17]
```diff
@@ -63,32 +67,25 @@ router.post(
 /** ---------------------------------------------------------------------------
  * GET /document
```
Comment (leite08) [inline api/app/src/routes/medical/document.ts:68]: Just noticed this is wrong. Once we decide on the name, could you fix it, please?

File: api/app/src/routes/medical/document.ts [hunk 18]
```diff
@@ -27,8 +31,8 @@ router.get(
     const facilityId = getFromQueryOrFail("facilityId", req);
     const forceQuery = getFromQuery("force-query", req);
 
-    const documents = await getDocuments({ cxId, patientId });
-    const documentsDTO = documents.map(toDTO);
+    const documents = await getDocuments(patientId);
```
Comment (leite08) [inline api/app/src/routes/medical/document.ts:34]: Why aren't we calling our business logic/command here? Doing like this is tying our API to Medplum code, whereas the command would call `/external` and convert to our own shape/type.

If we're to leave as is, I think we should remove code we're not using anymore.

File: api/app/src/routes/medical/dtos/documentDTO.ts [hunk 19]
```diff
@@ -1,6 +1,8 @@
-import { DocumentReference } from "../../../domain/medical/document-reference";
+import { DocumentReference, DocumentReferenceContent } from "@medplum/fhirtypes";
```
Comment (leite08) [inline api/app/src/routes/medical/dtos/documentDTO.ts:1]: ditto Medplum's code becoming a dependency across our code (ideally, this would be an import to our domain types).

File: api/app/src/routes/medical/organization.ts [hunk 20]
```diff
@@ -35,6 +37,10 @@ router.post(
     const createOrg: OrganizationCreateCmd = { cxId, ...data };
     const org = await createOrganization(createOrg);
 
+    // temp solution until we migrate to FHIR
+    const fhirOrg = toFHIR(org);
+    await upsertOrgToFHIRServer(fhirOrg);
```
Comment (leite08) [inline api/app/src/routes/medical/organization.ts:42]: Just relating to the comments on error handling of these: if we want to abstract away errors to FHIR server from our users, we could make these asynchronous. 

Here and other similar places.

Moot comment if we want to expose FHIR server related errors.

File: infra/lib/env-config.ts [hunk 21]
```diff
@@ -23,6 +23,7 @@ export type EnvConfig = {
   usageReportUrl?: string;
   fhirServerUrl?: string;
   systemRootOID: string;
+  medicalDocumentsBucketName?: string;
```
Comment (leite08) [inline infra/lib/env-config.ts:26]: This is optional even though we're referencing it on the API as required.

File: utils/src/fhir-migration.ts [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+function getEnvVarOrFail(varName: string): string {
+  const value = getEnvVar(varName);
+  if (!value || value.trim().length < 1) {
+    throw new Error(`Missing ${varName} env var`);
+  }
+  return value;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const toFHIROrg = (org: any): FHIROrganization => {
```
Comment (Orta21) [inline utils/src/fhir-migration.ts:66]: This disable doesnt work in the github build, the results from sequel arent typed so what would be the best approach to fixing this type issue besides make the file a js file

**Review by leite08 — CHANGES_REQUESTED** [review]
In addition to those comments:
- we not removing now obsolete code, like `models/medical/document-reference` and its table, and `command/medical/document/create-or-update` and `get-documents`; any particular reason or an oversight?

File: utils/src/fhir-migration.ts (hunk 22 above)
Comment (leite08) [reply utils/src/fhir-migration.ts:66]: I can't tell what the issue is by your comment, can you point to the failed job?

It does work to build local, right?

File: api/app/package.json [hunk 23]
```diff
@@ -41,6 +43,7 @@
     "lodash": "^4.17.21",
     "mime-types": "^2.1.35",
     "nanoid": "^3.3.4",
+    "node-fetch": "^3.3.1",
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply api/app/package.json:46]: Good catch - I think this is a left-over, fetch is included in node 18 by default (which is the main reason the ver was bumped)

File: api/app/src/external/commonwell/document/document-query.ts [hunk 24]
```diff
@@ -90,23 +104,83 @@ async function internalGetDocuments({
...
-              type: d.content.type,
-              status: d.content.status,
-              location: d.content.location,
-              indexed: d.content.indexed,
-              mimeType: d.content.mimeType,
-              size: d.content.size, // bytes
-              raw: d,
-            }
+        d.id && d.content && d.content.location
+          ? { id: d.id, content: { location: d.content.location, ...d.content } }
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply api/app/src/external/commonwell/document/document-query.ts:110]: Last I recall we'll want to use master ID

File: utils/src/fhir-migration.ts (hunk 22 above)
Comment (Orta21) [reply utils/src/fhir-migration.ts:66]: Yea it works locally this is the error https://github.com/metriport/metriport/actions/runs/4701736182/jobs/8338150898?pr=223

File: api/app/src/external/commonwell/document/shared.ts [hunk 25]
```diff
@@ -1,39 +1,84 @@
...
+    date: doc.content?.indexed,
+    status: "current",
+    type: doc.content?.type,
+    subject: {
+      reference: `Patient/${patient.id}`,
+      type: "Patient",
+    },
+    author: [
+      {
+        reference: `#${organization.id}`,
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply api/app/src/external/commonwell/document/shared.ts:56]: This is so it points to the Org resource in the `contained` field -> was necessary for responding to CW with no errors.

Probably not necessary here, but could leave as-is for consistency.

File: api/app/src/external/commonwell/document/shared.ts [hunk 26]
```diff
@@ -1,39 +1,84 @@
...
+        value: id.value,
+        // question
+        // use: id.use,
+      };
+    }),
+    date: doc.content?.indexed,
+    status: "current",
+    type: doc.content?.type,
+    subject: {
+      reference: `Patient/${patient.id}`,
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply api/app/src/external/commonwell/document/shared.ts:51]: It's a good point.

However, I'm wondering if we should make our internal IDs the primary ids for searchability, but then put the external IDs in extensions so we have them for reference?

File: api/app/src/external/commonwell/document/document-query.ts [hunk 27]
```diff
@@ -1,40 +1,53 @@
...
-    // update CW's entries, we replace them
-    await DocumentReferenceModel.destroy({
-      where: {
-        cxId: patient.cxId,
-        patientId: patient.id,
-        source: MedicalDataSource.COMMONWELL,
-      },
+    if (Config.isSandbox()) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:37]: Cant do upstream because we need to call this to complete polling, but ill put it at the top of the try catch

`await updateDocQueryStatus({ patient, status: "completed" });`

File: api/app/src/external/commonwell/document/document-query.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:110]: I had it in there it might of got erased during a merge conflict. Added it back

File: utils/src/fhir-migration.ts (hunk 22 above)
Comment (leite08) [reply utils/src/fhir-migration.ts:66]: Additionally, about TS referring to stuff from parent folder (mimicking Node's): https://github.com/microsoft/TypeScript/issues/30124

File: api/app/src/external/fhir/organization/upsert-organization.ts (hunk 16 above)
Comment (Orta21) [reply api/app/src/external/fhir/organization/upsert-organization.ts:12]: I changed all of these to not have a try catch so the asynchandler handles it

File: api/app/src/external/commonwell/document/shared.ts [hunk 28]
```diff
@@ -1,39 +1,84 @@
...
+import { MedicalDataSource } from "../..";
 
 // TODO #340 When we fix tsconfig on CW SDK we can remove the `Required` for `id`
-export type DocumentWithLocation = Required<Pick<Document, "id">> &
-  Omit<DocumentContent, "location"> &
-  Required<Pick<DocumentContent, "location">> & {
-    fileName: string;
-    raw?: unknown;
-  };
+export type DocumentWithFilename = Document & {
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/shared.ts:10]: I thought because we had it on the `Document` it would infer that it is required? Will this be fixed by tsconfig if not then ill make it required

File: api/app/src/routes/medical/document.ts (hunk 18 above)
Comment (Orta21) [reply api/app/src/routes/medical/document.ts:34]: Gotcha i was just under the impression that logic that lived within the commands was to interact with the models we have for the sql DB. 

If i understand correctly i should move just the getDocuments from medplum into a command and then reference that? After which the DTO still remains within the routes since thats where it lives

File: api/app/src/routes/medical/dtos/documentDTO.ts (hunk 19 above)
Comment (Orta21) [reply api/app/src/routes/medical/dtos/documentDTO.ts:1]: My understanding is we are using medplum fhirtypes as the defacto for setting FHIRtypes so we dont have to do that ourselves. If we want to not be reliant on them then i can work on recreating those types locally.

File: api/app/src/routes/medical/document.ts (hunk 18 above)
Comment (Orta21) [reply api/app/src/routes/medical/document.ts:34]: As per standup conversation, we will stick with medplum for now

File: api/app/src/routes/medical/dtos/documentDTO.ts (hunk 19 above)
Comment (Orta21) [reply api/app/src/routes/medical/dtos/documentDTO.ts:1]: As per standup conversation, we will stick with medplum for now

File: api/app/src/external/commonwell/document/shared.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (leite08) [reply api/app/src/external/commonwell/document/shared.ts:10]: Based on the comment there, I think we were not exposing the type properly from the package, so we had to add the `<Required>` there.

Also, the previous type was flattening the document, so there was no `content` on `DocumentWithLocation`. I don't know if that would still make sense with the changes on this PR, but previously I thought it made sense to simplify mapping to the local type of document references on the API (slightly diff from the one from CW SDK).

File: api/app/src/external/commonwell/document/document-query.ts [hunk 29]
```diff
@@ -1,40 +1,53 @@
...
   try {
-    const cwDocuments = await internalGetDocuments({ patient, facilityId });
+    const { organization, facility } = await getPatientData(patient, facilityId);
 
-    const documents = cwDocuments.map(toDomain(patient));
+    const cwDocuments = await internalGetDocuments({ patient, organization, facility });
 
-    // Commonwell does not consistently return the same ID for a given document, so we don't
-    // update CW's entries, we replace them
-    await DocumentReferenceModel.destroy({
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:29]: Spoke to Dima about this and we are going to allow for duplicates for now and he has also opened a case with commonwell in regards to the documents references having different ids

File: api/app/src/external/commonwell/document/shared.ts [hunk 30]
```diff
@@ -1,39 +1,84 @@
...
+  const id = doc.id?.replace("urn:uuid:", "");
+
+  return {
+    id: id,
+    resourceType: "DocumentReference",
+    contained: [
+      {
+        resourceType: "Organization",
+        id: organization.id,
+        name: organization.data.name,
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/shared.ts:28]: Were going to store the docs org and patient as an aside

File: api/app/src/external/commonwell/document/shared.ts [hunk 31]
```diff
@@ -1,39 +1,84 @@
...
+      },
+    ],
+    extension: [
+      {
+        valueReference: {
+          reference: MedicalDataSource.COMMONWELL,
+        },
+      },
+    ],
+    context: doc.content?.context,
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/shared.ts:79]: I believe we are storing all the data the comes back from the document

File: api/app/src/external/commonwell/document/document-query.ts [hunk 32]
```diff
@@ -90,23 +104,83 @@ async function internalGetDocuments({
...
+
+        await downloadDocument({
+          cxId: patient.cxId,
+          patientId: patient.id,
+          facilityId: facilityId,
+          location: doc.content?.location ? doc.content.location : "",
+          stream: writeStream,
+        });
+
+        const data = await promise;
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:155]: This is wrapped in a promise.allsettled, or do you mean to also wrap ` const data = await promise;` in all settled as well

File: api/app/src/external/commonwell/document/shared.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/shared.ts:10]: I merged your PR and id is now being flagged as required, as for flattening it makes more sense to leave content as is because thats how its being added when we convert to FHIR

File: api/app/src/external/commonwell/document/document-query.ts (hunk 32 above)
[the remark this answers is not in the document]
Comment (leite08) [reply api/app/src/external/commonwell/document/document-query.ts]: The `Promise.allSettled` is supporting parallel processing of files. But each file's code is processed in sequence.

In the processing of a file, we await the download of the file at line 147. Sure, its being downloaded to the same stream that's being used by S3 to upload the file, so maybe its working as supposed (a byte is written on the stream, a byte is read and taken out of that stream).

My concern is that there's no explicit "piping" of data, like its done [here](https://stackoverflow.com/a/73332454/2099911), so I can't tell if we're first downloading all file on line 147 for only later upload the contents to S3 on line 155, or if it happens seamlessly like when we use `.pipe()`.

Not sure if we could monitor the stream's buffer usage to see if the file is being download entirely before being uploaded, but NABD and we can just move forward if the current code is working.

[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts]: Definitely a valid point, my assumption was that this worked similar to when we were attaching it to the express res. If you go down the chain of functions from download document it gets to this point `await pipeline(response.data, outputStream);` where it seems like its piping to the s3 write stream

[the remark this answers is not in the document]
Comment (leite08) [reply api/app/src/external/commonwell/document/document-query.ts]: > If you go down the chain of functions from download document it gets to this point await pipeline(response.data, outputStream); where it seems like its piping to the s3 write stream

Yeah, 💯! But that's writing on the stream... _I'm assuming_ S3 will start uploading as bytes become available on that stream, but not sure as I'm more familiar with awaiting on `.pipe`, not on a download that puts on a stream that has other asynchronous function reading from.

I'm fine with moving forward with the current version, if we start noticing slow response times to download files we can come back at this and add code to "instrument"/monitor the stream/buffer usage to see if it is storing the whole file before uploading or just a small amount of bytes before they get uploaded (the piping). 👍

File: api/app/src/models/db.ts [hunk 33]
```diff
@@ -24,7 +23,6 @@ const models: ModelSetup[] = [
   PatientModel.setup,
   MAPIAccess.setup,
   CustomerSequenceModel.setup,
-  DocumentReferenceModel.setup,
```
Comment (leite08) [inline api/app/src/models/db.ts:27]: I see we're not deleting the table now, which I think makes sense. Maybe create a ticket to remove it later?

File: api/app/src/external/commonwell/document/document-query.ts [hunk 34]
```diff
@@ -90,27 +105,108 @@ async function internalGetDocuments({
...
-              indexed: d.content.indexed,
-              mimeType: d.content.mimeType,
-              size: d.content.size, // bytes
-              raw: d,
-            }
-          : []
-      )
+  const documents: Document[] = docs.entry
+    ? docs.entry.flatMap(d => {
+        if (d.content.size && d.content.size === 0) {
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:110]: Good catch i removed the first check

File: api/app/src/external/commonwell/document/document-query.ts [hunk 35]
```diff
@@ -90,27 +105,108 @@ async function internalGetDocuments({
...
+    documents.map(async doc => {
+      try {
+        if (doc.content?.masterIdentifier?.value) {
+          const { writeStream, promise } = uploadStream(doc.content?.masterIdentifier?.value);
+
+          await downloadDocument({
+            cxId: patient.cxId,
+            patientId: patient.id,
+            facilityId: facilityId,
+            location: doc.content?.location ? doc.content.location : "",
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/external/commonwell/document/document-query.ts:180]: I dont think so so ill add this to the if check

File: api/app/src/models/db.ts (hunk 33 above)
Comment (Orta21) [reply api/app/src/models/db.ts:27]: https://github.com/metriport/metriport-internal/issues/605

File: api/app/src/routes/medical/organization.ts [hunk 36]
```diff
@@ -66,6 +72,10 @@ router.put(
     };
     const org = await updateOrganization(updateCmd);
 
+    // temp solution until we migrate to FHIR
+    const fhirOrg = toFHIR(org);
+    await upsertOrgToFHIRServer(fhirOrg);
```
[the remark this answers is not in the document]
Comment (Orta21) [reply api/app/src/routes/medical/organization.ts:77]: https://github.com/metriport/metriport-internal/issues/606

File: utils/src/fhir-migration.ts (hunk 22 above)
Comment (Orta21) [reply utils/src/fhir-migration.ts:66]: Yea theres no more issue with this

File: api/app/src/routes/medical/dtos/documentDTO.ts [hunk 37]
```diff
@@ -19,6 +20,14 @@ export function toDTO(docs: DocumentReference[] | undefined): DocumentReferenceD
       if (doc && doc.id && doc.content) {
         const hasAttachment = doc.content[0];
 
+        if (doc.content.length > 1) {
+          capture.message("Doc contains more than one content item", {
+            extra: {
+              content_length: doc.content.length,
+            },
```
Comment (leite08) [inline api/app/src/routes/medical/dtos/documentDTO.ts:27]: Can we add the doc identification as well?

**Review by leite08 — APPROVED** [review]
LGTM! Just left a small comment.

File: api/app/src/external/commonwell/document/document-download.ts [hunk 38]
```diff
@@ -41,4 +43,33 @@ export async function downloadDocument({
...
+  //       extra: {
+  //         context: `cw.retrieveDocument`,
+  //         cwReference: commonWell.lastReferenceHeader,
+  //         ...(err instanceof CommonwellError ? err.additionalInfo : undefined),
+  //       },
+  //     });
+  //     done = true;
+  //     throw new NotFoundError("Document not found");
+  //   }
+  // }
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/document-download.ts:74]: leftover?

File: api/app/src/external/commonwell/document/document-query.ts [hunk 39]
```diff
@@ -1,40 +1,54 @@
...
+    const docsS3Refs = await dowloadAndUploadDocsToS3({
+      patient,
+      facilityId,
+      documents: cwDocuments,
     });
 
-    return await createOrUpdate(patient, documents);
+    for (const doc of docsS3Refs) {
+      const FHIRDocRef = toFHIRDocRef(doc, organization, patient);
+      await upsertDocumentToFHIRServer(FHIRDocRef);
```
Comment (Goncharo) [inline api/app/src/external/commonwell/document/document-query.ts:50]: If we're not going to do the above suggestion now - can we not do a bulk upsert?

300 docs === 300 db calls, which in theory could be done in 1

**Review by Goncharo — CHANGES_REQUESTED** [review]
Couple points to consider - also reminder to update the docs

File: docs/medical-api/api-reference/document/get-document.mdx [hunk 40]
```diff
@@ -1,28 +1,20 @@
...
+description: "Fetches the document and sends a presigned URL"
+api: "GET /medical/v1/document/downloadUrl"
 ---
 
-## Query Params
-
-<ParamField query="patientId" type="string" required>
-  The ID of the Patient for which to list available Documents.
-</ParamField>
+This endpoint returns a URL which you can use to download the specified document using the file name provided from the [List Documents endpoint](/medical-api/api-reference/document/list-documents)
```
Comment (Goncharo) [inline docs/medical-api/api-reference/document/get-document.mdx:7]: hyper nit: missing `.`

File: docs/medical-api/api-reference/document/get-document.mdx [hunk 41]
```diff
@@ -1,28 +1,20 @@
...
+<ParamField query="fileName" type="string" required>
+  The file name of the document.
 </ParamField>
 
 ## Response
 
-<ResponseField type="blob">
-  The downloaded Document contents.
-</ResponseField>
+<ResponseField type="string">The presigned url.</ResponseField>
```
Comment (Goncharo) [inline docs/medical-api/api-reference/document/get-document.mdx:17]: nit: for consistency, `URL`

File: docs/medical-api/api-reference/document/get-document.mdx (hunk 40 above)
Comment (Goncharo) [reply docs/medical-api/api-reference/document/get-document.mdx:7]: Should we also add a warning that the presigned URL will only be valid for 1 minute (or whatever the duration was)?

File: docs/medical-api/api-reference/document/get-document.mdx [hunk 42]
```diff
@@ -1,28 +1,20 @@
 ---
 title: "Get Document"
-description: "Downloads the Document contents of the specified URL."
-api: "GET /medical/v1/document/download"
+description: "Fetches the document and sends a presigned URL"
```
Comment (Goncharo) [inline docs/medical-api/api-reference/document/get-document.mdx:3]: Is this true? We don't fetch the doc here, just return a URL.

Would this be more accurate? `Gets a presigned URL for downloading the specified document.`

File: docs/medical-api/api-reference/document/get-document.mdx [hunk 43]
```diff
@@ -1,28 +1,20 @@
 ---
 title: "Get Document"
```
Comment (Goncharo) [inline docs/medical-api/api-reference/document/get-document.mdx:2]: Perhaps `Get Document URL`?

**Review by Goncharo — APPROVED** [review]
Changes LGTM - approval contingent on comments

File: api/app/src/routes/medical/document.ts (hunk 13 above)
Comment (leite08) [reply api/app/src/routes/medical/document.ts:72]: yeah, but `url` would point to `/document`, so the browser deals with the redirect and download :)

File: api/app/src/external/commonwell/document/document-query.ts (hunk 39 above)
Comment (leite08) [reply api/app/src/external/commonwell/document/document-query.ts:50]: OTOH, if one fails we might fail all... not sure if the FHIR server would treat a batch upsert as a logical transaction or not. Just something to consider, I don't know the answer to that.

File: docs/medical-api/api-reference/document/get-document.mdx (hunk 40 above)
Comment (leite08) [reply docs/medical-api/api-reference/document/get-document.mdx:7]: Good point. I suggested we don't return a URL, but return a redirect to the S3 presigned URL... in which case the doc can be updated to state there's a redirect on the response. OTOH, if we decide to keep returning the URL, I think we might need to increase the expiring duration from 20s to something close to 1min - who knows, they might want to have a queue between getting the URL and actual downloading the file.

But I would go with the redirect instead of returning a URL. 😄

Comment (leite08) [reply docs/medical-api/api-reference/document/get-document.mdx:7]: Let's leave as is and try to build the redirect later on https://github.com/metriport/metriport-internal/issues/609

Comment (leite08) [reply docs/medical-api/api-reference/document/get-document.mdx:7]: So, back to Dima's comments:
- hyper nit: missing .
- add a warning that the presigned URL will only be valid for 1 minute (or whatever the duration was)
